### PR TITLE
[claude-generated] fix(utils): prevent remove_think_tags from truncating responses with embedded <think> tags

### LINKED ISSUE #2895: [Bug]:Responses are truncated when retrieved chunks contain <think> tags
### Do you need to file an issue?

- [ ] I have searched the existing issues and this bug is not already filed.
- [x] I believe this is a legitimate bug, not just a question or feature request.

### Describe the bug

I’ve noticed that when a retrieved chunk contains `xxx<think>xxx</think>xxx`, the frontend response is being truncated. It seems the logic intended to strip the reasoning content is accidentally cutting off the legitimate AI output as well.

### Steps to reproduce

_No response_

### Expected Behavior

_No response_

### LightRAG Config Used

# Paste your config here

### Logs and screenshots

_No response_

### Additional Information

- LightRAG Version:
- Operating System:
- Python Version:
- Related Issues:

### PR DESCRIPTION
**Pull request by sjhddh**
## Summary

- Fixes the `remove_think_tags` regex in `lightrag/utils.py` which caused false truncation of LLM responses when retrieved chunks contained `<think>` tags (issue #2895)
- The old regex `^(<think>.*?</think>|.*</think>)` with `re.DOTALL` would greedily match everything from the start of the string to the last `</think>`, discarding legitimate content before embedded `<think>` blocks
- The new implementation handles two cases separately: (1) orphaned `</think>` at the start (streaming), and (2) complete `<think>...</think>` blocks anywhere in the text
- Added 10 unit tests covering normal usage, the bug scenario, edge cases, and multiline content

Closes #2895

## Test plan

- [x] All 10 new tests in `tests/test_remove_think_tags.py` pass
- [x] Verified old regex fails on the reported bug case (mid-text `<think>` tags)
- [x] Verified new regex correctly handles: standard think blocks at start, orphaned `</think>` prefix, no think tags, think tags in middle of text, multiple blocks, multiline blocks, empty blocks

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by danielaskdd** [comment]
@codex  review

File: lightrag/utils.py [hunk 1]
```diff
@@ -1946,11 +1946,19 @@ async def update_chunk_cache_list(
...
+    """Remove <think>...</think> tags and their content from the text.
+
+    Handles two cases:
+    1. Complete <think>...</think> blocks anywhere in the text.
+    2. Orphaned </think> at the very start (e.g., from streaming that begins
+       mid-think-block), removing everything before and including it.
+    """
+    # First, remove orphaned </think> prefix (content before first </think>
+    # when there is no preceding <think> tag)
+    text = re.sub(r"^[^<]*</think>", "", text, flags=re.DOTALL)
```
[the remark this answers is not in the document]
Comment (sjhddh) [reply lightrag/utils.py]: Good catch! Fixed in 28421fc — changed the orphan regex from `^[^<]*</think>` to `^((?!<think>).)*?</think>`, which uses a negative lookahead to skip `<think>` specifically rather than blocking all `<` characters. Added tests for the `2 < 3` and HTML-in-reasoning cases.

**Comment by danielaskdd** [comment]
@codex review

File: tests/test_remove_think_tags.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+Covers the fix for issue #2895: responses truncated when retrieved chunks
+contain <think> tags.
+"""
+
+import re
+
+import pytest
+
+
+def remove_think_tags(text: str) -> str:
```
[the remark this answers is not in the document]
Comment (sjhddh) [reply tests/test_remove_think_tags.py]: Addressed in d6019bd — tests now `try: from lightrag.utils import remove_think_tags` first, falling back to a local copy only when deps are missing. In CI (where lightrag is installed), the real implementation is always tested.

**Comment by danielaskdd** [comment]
@codex review

File: lightrag/utils.py [hunk 3]
```diff
@@ -1946,11 +1946,19 @@ async def update_chunk_cache_list(
...
+    """Remove <think>...</think> tags and their content from the text.
+
+    Handles two cases:
+    1. Complete <think>...</think> blocks anywhere in the text.
+    2. Orphaned </think> at the very start (e.g., from streaming that begins
+       mid-think-block), removing everything before and including it.
+    """
+    # First, remove orphaned </think> prefix (content before first </think>
+    # when there is no preceding <think> tag)
+    text = re.sub(r"^((?!<think>).)*?</think>", "", text, flags=re.DOTALL)
```
[the remark this answers is not in the document]
Comment (sjhddh) [reply lightrag/utils.py:1958]: This is an inherent ambiguity that regex alone cannot resolve. Consider `"reasoning mentions <think> marker</think>The answer"` — it is indistinguishable from normal text followed by a complete `<think>...</think>` block (which is the far more common case and the original bug #2895).

The current behavior treats `<think>` as an opening tag, which is correct for the real-world scenario: retrieved chunks containing `<think>...</think>` pairs embedded in response text. The edge case where orphaned streaming content itself contains the literal string `<think>` is vanishingly unlikely (it would mean an LLM's reasoning references its own control tokens as plain text within a mid-stream cut).

Resolving this perfectly would require streaming-level context (knowing whether a `<think>` was opened in a prior chunk), which is outside the scope of this stateless helper. The tradeoff here is correct for production usage.

**Comment by danielaskdd** [comment]
@codex review
